src/java/com/rapleaf/jack/queries/GenericQueryBuilder.java

```diff
@@ -0,0 +1,176 @@
+package com.rapleaf.jack.queries;
+
+import java.io.IOException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLRecoverableException;
+import java.util.Set;
+
+import com.rapleaf.jack.BaseDatabaseConnection;
+import com.rapleaf.jack.Column;
+import com.rapleaf.jack.Table;
+import com.rapleaf.jack.queries.where_operators.IWhereOperator;
+
+public class GenericQueryBuilder {
+  private static int MAX_CONNECTION_RETRIES = 1;
+
+  private final GenericQuery genericQuery;
+  private final BaseDatabaseConnection dbConnection;
+
+  public GenericQueryBuilder(BaseDatabaseConnection dbConnection, GenericQuery genericQuery) {
+    this.dbConnection = dbConnection;
+    this.genericQuery = genericQuery;
+  }
+
+  public JoinConditionBuilder leftJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.LEFT_JOIN, table);
+  }
+
+  public JoinConditionBuilder rightJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.RIGHT_JOIN, table);
+  }
+
+  public JoinConditionBuilder innerJoin(Table table) {
+    return new JoinConditionBuilder(this, JoinType.INNER_JOIN, table);
+  }
+
+  GenericQueryBuilder addJoinCondition(JoinCondition joinCondition) {
+    genericQuery.addJoinCondition(joinCondition);
+    return this;
+  }
+
+  public GenericQueryBuilder where(Column column, IWhereOperator operator) {
```

Comment:
Any `WHERE` clause with a null logic other than the first one will default to `AND`. I know this is not great.

Putting everything into only one `WHERE` clause can solve both problems. I can think of two approaches:
1. The OOP way: `where(a.and(b.or(c)))`;
2. The LISP way: `where(AND(a, OR(b, c)))`.

In both approaches, `a = User.ID.equalsTo(123L)`, `b = User.HANDLER.equalsTo("abc")` and `c = User.BIO.equalsTo("abc")`, in which `equalsTo` is a method called on a `Column` and returns a where constraint.